internal/pkg/agent/application/actions/handlers/handler_action_upgrade.go

```diff
@@ -30,13 +33,38 @@ func NewUpgrade(log *logger.Logger, coord *coordinator.Coordinator) *Upgrade {
 	}
 }
 
-// Handle handles UPGRADE action.
-func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, _ acker.Acker) error {
+// Handle handles UPGRADE action.  Returns immediately and the actual
+// upgrade happens asynchronously.  This allows for downloads to
+// happen without blocking updates.  If a second UPGRADE action is
+// received before the first is completed, then an error is returned.
+func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, ack acker.Acker) error {
 	h.log.Debugf("handlerUpgrade: action '%+v' received", a)
 	action, ok := a.(*fleetapi.ActionUpgrade)
 	if !ok {
 		return fmt.Errorf("invalid type, expected ActionUpgrade and received %T", a)
 	}
-
-	return h.coord.Upgrade(ctx, action.Version, action.SourceURI, action)
+	go func() {
+		h.m.Lock()
+		if h.bkgAction != nil {
+			h.log.Infof("upgrade to version %s already running in background", h.bkgAction.Version)
+			h.m.Unlock()
+			return
+		}
+		h.bkgAction = action
+		h.m.Unlock()
+		h.log.Infof("starting upgrade to version %s in background", action.Version)
+		if err := h.coord.Upgrade(ctx, action.Version, action.SourceURI, action); err != nil {
+			h.log.Errorf("upgrade to version %s failed: %v", action.Version, err)
+			if err := ack.Ack(ctx, action); err != nil {
+				h.log.Errorf("ack of failed upgrade failed: %v", err)
+			}
+			if err := ack.Commit(ctx); err != nil {
+				h.log.Errorf("commit of ack for failed upgrade failed: %v", err)
+			}
```

Comment:
I don't believe these should be here. The original code on the left side of the PR shows that it only called the `h.coord.Upgrade`. The ack and commit is handled by the upgrade manager inside of the coordinator.

Reply:
If these aren't here, then if there is an error in the upgrade the fleet server never gets updated status.  I think this is because in the dispatcher it calls ack.Commit, but because we are in our own go routine it doesn't have our ack yet to commit.  If the upgrade is successful, we send the ack on reboot.

Reply:
@blakerouse any thoughts on making sure the ACK gets delivered?

Reply:
Hmm so I assume that previous code was incorrect?

Reply:
from zoom conversation,  the ack has to happen here because we are in our own go function.  Previously it was handled by the dispatcher, but now the dispatcher won't see any errors.